### ISSUE TITLE
Unable to update Slack alert_configurations via Oauth integration 

### ISSUE THREAD

**Issue by TheQueenIsDead** [mongodb/terraform-provider-mongodbatlas#1074]
### Terraform CLI and Terraform MongoDB Atlas Provider Version

```hcl
$ terraform version
Terraform v1.3.9
on linux_amd64
+ provider registry.terraform.io/hashicorp/vault v3.12.0
+ provider registry.terraform.io/mongodb/mongodbatlas v1.8.0
```

### Terraform Configuration File 

```hcl
resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned" {
  project_id = mongodbatlas_project.this.id
  event_type = "OUTSIDE_METRIC_THRESHOLD"
  enabled    = true

  metric_threshold_config {
    metric_name = "QUERY_TARGETING_SCANNED_OBJECTS_PER_RETURNED"
    mode        = "AVERAGE"
    operator    = "GREATER_THAN"
    threshold   = 1000
    units       = "RAW"
  }

  # Alert to slack for development environment
  dynamic "notification" {
    for_each = terraform.workspace == "development" ? ["alert"] : []
    content {
      channel_name  = var.alert_slack_channel
      delay_min     = 0
      email_enabled = false
      interval_min  = 30
      roles         = []
      type_name     = "SLACK"
    }
  }

  # Send a default email in prod
  dynamic "notification" {
    for_each = terraform.workspace == "production" ? ["alert"] : []
    content {
      delay_min     = 5
      email_enabled = true
      interval_min  = 60
      roles         = ["GROUP_OWNER"]
      type_name     = "GROUP"
    }
  }
}
```

### Steps to Reproduce
1. Setup [Slack integration for a project](https://www.mongodb.com/docs/cloud-manager/tutorial/slack-integration/).
2. Use the [Alert Configurations](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/data-sources/alert_configurations) resource to import default email alerts for a project.
3. Ensure that a plan yields no differences.
4. Edit the alert via the UI to use Slack instead of email (See picture)
![image](https://user-images.githubusercontent.com/30706552/220208598-3d581608-56bf-4219-9a19-a5560871411c.png)
5. Change an attribute in the notification Terraform like `delay_min`
6. Apply the change.
7. Receive an error.

### Expected Behavior

Terraform should have been able to execute an update to a slack notification without passing in a token. In the same way that the UI populates that for the user, it would be super if the backend was capable of retrieving the token itself (Especially with the new Oauth)

Or perhaps, I would have been able to retrieve the token via a data source, and pass it in. However, `SLACK` is an unsupported type in the third party integration data source:
https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/data-sources/third_party_integration#type

### Actual Behavior

Terraform returned an error as the token is expected. 
```hcl
mongodbatlas_alert_configuration.query_targeting_scanned_objects_per_returned: Modifying... [id=...]
│ Error: error updating Alert Configuration information: PUT https://cloud.mongodb.com/api/atlas/v1.0/groups/5b...b7/alertConfigs/63...41: 400 (request "MISSING_ATTRIBUTE") The required attribute apiToken was not specified.
│ 
│   with mongodbatlas_alert_configuration.query_targeting_scanned_per_returned,
│   on alert_configuration.tf line 53, in resource "mongodbatlas_alert_configuration" "query_targeting_scanned_per_returned":
│   53: resource "mongodbatlas_alert_configuration" "query_targeting_scanned_per_returned" {
│ 
╵
```

Would be happy to have a workaround suggested for this in the meantime, however I can not see one outside of editing the alert via the UI, which is a shame as I would like to move most of our alerts to include slack.

Thanks!

**Comment by martinstibbe** [comment]
@TheQueenIsDead The apiToken value is a required value if sending an update via Terraform to Atlas API.
The datasource route would not work as apiToken value (and many others from other Third Party Integration ) are obfuscated coming back from Atlas API

**Comment by TheQueenIsDead** [comment]
Hi, I'm not so certain about that. As mentioned above, the UI seems to be capable of setting up a Slack message. When I look at the PUT request that is sent by the browser, the payload looks something like this (PII redacted):

```json
# PUT https://cloud.mongodb.com/activity/alertConfigs/5b<REDACTED>b7/63<REDACTED>eb
{
  "_t": "HOST_METRIC",
  "et": "OUTSIDE_METRIC_THRESHOLD",
  "cid": "5b<REDACTED>b7",
  "enabled": true,
  "upd": "2023-02-21T23:46:36Z",
  "ac": "GROUP",
  "forAllGroups": false,
  "groupType": "NDS",
  "tags": [],
  "matchers": [],
  "cre": "2023-02-20T23:45:03Z",
  "groups": [],
  "id": "63<REDACTED>eb",
  "mt": {
    "metric": "QUERY_TARGETING_SCANNED_OBJECTS_PER_RETURNED",
    "mode": "AVERAGE",
    "units": "RAW",
    "op": "GREATER_THAN",
    "threshold": 1000
  },
  "notify": [
    {
      "_t": "SLACK",
      "interval": 60,
      "delay": 0,
      "integration": {
        "_t": "SLACK",
        "apiToken": "*************************************************************************<REDACTED>",
        "channelName": "#monitoring-dev",
        "teamName": "<REDACTED>"
      }
    }
  ],
  "orgId": null,
  "source": "UI"
}

```

I captured this event, re-adjusted the alert to use email, and then executed the request as a curl command, where it succeeded in re-applying the Slack notification.
It would seem that sending the apiToken, _even as a string of literal stars_, yields a successful PUT request, which is super interesting.

I thought the UI might use the same API as the Terraform provider, but I see that the above request reaches out to `https://cloud.mongodb.com/activity/alertConfigs/5b<REDACTED>b7/63<REDACTED>eb` whereas the [API seems to specify](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Alert-Configurations/operation/updateAlertConfiguration) `https://cloud.mongodb.com/api/atlas/v1.0/groups/{groupId}/alertConfigs/{alertConfigId}`

I suppose the concrete thing I'm asking is: If the UI endpoint can not be used, there is no backend API equivalent, does this mean that I need to configure a legacy webhook integration with Slack in order to get the token and be unblocked? 

It would be a shame to learn that no planning to accommodate Slack alerts via API was planned as the Oauth integration moved ahead.

Any workarounds appreciated :pray:

**Comment by themantissa** [comment]
@TheQueenIsDead really impressive investigation here!   As you've noticed the UI does not use the Admin API, they are actually different.  So the UI behavior is not always the same as the Admin API.  But you bring up a good point - let me talk with the Product Managers in these areas and see what our options might be here.

**Comment by TheQueenIsDead** [comment]
Super, thanks @themantissa ! This is certainly more of a feature request than a bug, as I'm aware that the API would likely require a change to support this (Which is somewhat outside of the scope of this provider)

**Comment by Zuhairahmed** [comment]
@TheQueenIsDead just as an interim suggestion while MongoDB team continues to investigate this issue (internal ticket [CLOUDP-167701](https://jira.mongodb.org/browse/CLOUDP-167701)), have you considered storing Slack API token secret after it has been created into Vault (or another secrets manager). this way you can simply query Vault in your config file to pull Slack API token as an input whenever you want to use resource `mongodbatlas_alert_configuration`

**Comment by TheQueenIsDead** [comment]
Hi @Zuhairahmed , If I could retrieve the token that would absolutely be a fine interim solution. However, as noted above, the token is obfuscated on the UI, and there is no data provider for a slack `third_party_integration`.

If you could point me to where I can retrieve that token once setup, that would be super

**Comment by Zuhairahmed** [comment]
hi @TheQueenIsDead just as an update, team is working on updating resource to allow users to retrieve Slack API Token and will update documentation accordingly as well. will let you know when this has been completed, but likely will be still several months away. feel free to reach out in the interim if you need anything else here.

**Comment by TheQueenIsDead** [comment]
That's neat to hear, thanks.

in the meantime I've created a Slack app with incoming webhooks and changed to using the `WEBHOOK` type to send notifications.

An example, for those that may be in the same boat:

```terraform
# This is configured outside of the project via the UI
resource "mongodbatlas_third_party_integration" "slack_webhook" {
  project_id = mongodbatlas_project.this.id
  type       = "WEBHOOK"
  url        = var.project_alert_webhook_url
}

resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned" {
  project_id = mongodbatlas_project.this.id
  event_type = "OUTSIDE_METRIC_THRESHOLD"
  enabled    = true

  metric_threshold_config {
    metric_name = "QUERY_TARGETING_SCANNED_OBJECTS_PER_RETURNED"
    mode        = "AVERAGE"
    operator    = "GREATER_THAN"
    threshold   = 1000
    units       = "RAW"
  }

  # Alert to Slack
  notification {
    delay_min    = 1
    interval_min = 60
    type_name    = "WEBHOOK"
    webhook_url  = mongodbatlas_third_party_integration.slack_webhook.url
  }
}
```

**Comment by AgustinBettati** [comment]
Hi @TheQueenIsDead!
Wanted to let you know that a new attribute `notifier_id` has been added enabling third party notifications to be updated without having to provide original credentials https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1514. This is targeted for our next 1.12.2 release, I will reach out when it is available.

**Comment by AgustinBettati** [comment]
Hello,

v1.12.2 has been released with the support of the new `notifier_id` attribute. Closing this issue, feel free to reopen if you have any remaining concerns on this matter.

**Comment by TheQueenIsDead** [comment]
@AgustinBettati really neat to see that this use-case is being supported and worked on, thank you :-) 

Would you be able to confirm where the `notifier_id` is meant to be sourced from? I setup the following but encountered an issue:

```terraform

data "mongodbatlas_third_party_integration" "slack" {
  project_id = mongodbatlas_project.this.id
  type       = "SLACK"
}

resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned" {
  project_id = mongodbatlas_project.this.id
  event_type = "OUTSIDE_METRIC_THRESHOLD"
  enabled    = true

  metric_threshold_config {
    metric_name = "QUERY_TARGETING_SCANNED_OBJECTS_PER_RETURNED"
    mode        = "AVERAGE"
    operator    = "GREATER_THAN"
    threshold   = 1000
    units       = "RAW"
  }

  notification {
    email_enabled = true
    interval_min  = 60
    roles         = ["GROUP_OWNER"]
    type_name     = "GROUP"
  }

  # Alert to Slack
  notification {
    delay_min    = 1
    interval_min = 60
    type_name    = "SLACK"
    notifier_id  = data.mongodbatlas_third_party_integration.slack.id
  }
}
```
```
╷
│ Error: error updating Alert Configuration information: %s
│ 
│   with mongodbatlas_alert_configuration.query_targeting_scanned_objects_per_returned,
│   on alert_configuration.tf line 12, in resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned":
│   12: resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned" {
│ 
│ PUT https://cloud.mongodb.com/api/atlas/v1.0/groups/5b<XXX>b7/alertConfigs/64<XXX>3b: 400 (request "INVALID_JSON_ATTRIBUTE") Received JSON for the notifications.java.util.ArrayList[1].notifierId
│ attribute does not match expected format.
╵

```

I had a look in the Terraform state, and it doesn't look like a valid notifier ID is exposed in the data resource for the third party integration. The ID changes like so when I attempt to move from WEBHOOK to SLACK:
```terraform
 ~ notifier_id   = "642de7efe2ea8a3468551a2c" -> "cHJvamVjdF9pZA==:N<XXX>=:U0xBQ0s="
```

The ID seems to read "project_id:<ACTUAL_ID>:SLACK" when decoded, which isn't in keeping with the style of the webhook id

**Comment by nthienan** [comment]
experiencing the same issue as @TheQueenIsDead. Would appreciate any help!

**Comment by AgustinBettati** [comment]
Hello @TheQueenIsDead / @nthienan. Thanks for sharing these details. Let me describe how the use case of updating a third party alert configuration notification would look like:

- The `notifier_id` value is stored in the terraform state as this value is returned from the API. A command such as `terraform state show mongodbatlas_alert_configuration.<resource-name>` can be used to see the notifier id value of each notification.
- For the notification you want to update, you provide the corresponding `notifier_id` and this will enable a successful update without needing to provide original credentials.

Let me know if you encounter any issues, and I will also work on documenting these steps in our examples section to clear up future doubts on how to obtain this value.

**Comment by TheQueenIsDead** [comment]
Hi @AgustinBettati, thanks for the response!

I think I see where I misunderstood the change. I imagine that yes, an alert that has been pre-provisioned in the UI with the Slack Oauth integration can probably be imported and updated (I haven't tested this, but I'm sure your team has!)

It's been a while since I've paid this much thought, but I imagine that I had initially performed an import, found the issue, reported the bug, then moved on without attempting the next step, which in my head would be to pass the Slack token / `notifier_id` into the `alert_configuration` when the alert is being _provisioned_ (initial creation).

Terraform is most useful for this, as opposed to manually reconciling UI changes via import 😉

I postulated this in the initial bug report:
> Or perhaps, I would have been able to retrieve the token via a data source, and pass it in. However, SLACK is an unsupported type in the third party integration data source:
https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/data-sources/third_party_integration#type

Having seen that Slack can now be configured as a third party integration (It does not mention this in the Terraform docs explicitly, but makes reference to the `TYPE` in the API [docs](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Third-Party-Integrations/operation/createThirdPartyIntegration)), I would have assumed that the `id` returned by the `mongodbatlas_third_party_integration` `data` resource would be able to be used as the `notifier_id` in an `alert_configuration`.

Is that possible to achieve currently, or would it need more work within the provider? 

Let me know if I've got the wrong end of the stick or if you think there's a better way to go about what we're trying to achieve, hopefully this provides some insight into the true outcome I had imagined (Manually configured Slack OAuth integration in the UI, with the ability to reference that within Terraform in order to create new alert resources) 

Thanks again for your help thus far 😊

**Comment by TheQueenIsDead** [comment]
I suppose a workaround could be to find the value, then simply store it as a `local` variable, though integration with a `data` resource would be the ideal!

I'll give it a go when I get a spare chance

**Comment by nthienan** [comment]
@AgustinBettati my case is a bit different. It's about creating new alerts while Slack integration already created via web UI.

**Comment by nthienan** [comment]
@AgustinBettati my case is a bit different. It's about creating new alerts while Slack integration already created via web UI.

**Comment by TheQueenIsDead** [comment]
I've configured Slack integration on two alerts and ran a Terraform apply without changing the resource definitions. I can see that the notifier ID is different per alert despite ending up at the same Slack channel.

```tf
  # mongodbatlas_alert_configuration.query_targeting_scanned_objects_per_returned will be updated in-place
  ~ resource "mongodbatlas_alert_configuration" "query_targeting_scanned_objects_per_returned" {
      - notification {
          - channel_name  = "#monitoring-dev" -> null
          - delay_min     = 1 -> null
          - email_enabled = false -> null
          - interval_min  = 60 -> null
          - notifier_id   = "656cf1b240c74b06ab2e3a33" -> null
          - sms_enabled   = false -> null
          - type_name     = "SLACK" -> null
        }
    }

  # mongodbatlas_alert_configuration.query_targeting_scanned_per_returned will be updated in-place
  ~ resource "mongodbatlas_alert_configuration" "query_targeting_scanned_per_returned" {
      - notification {
          - channel_name  = "#monitoring-dev" -> null
          - delay_min     = 0 -> null
          - email_enabled = false -> null
          - interval_min  = 60 -> null
          - notifier_id   = "656cf22df9081e442331e2cd" -> null
          - sms_enabled   = false -> null
          - type_name     = "SLACK" -> null
        }
    }
```

So to get everything working we'd have to grab those values and include them with the alert configuration for each alert. That's pretty tedious, and does not lend itself well to people that use multiple workspaces (For example, if we had 20 alerts and 3 different terraform workspaces, we'd have to map to 60 different alert identifiers.

I don't imagine this lends itself nicely to creating the alert from scratch.

**Comment by AgustinBettati** [comment]
Hello @TheQueenIsDead  / @nthienan. Thanks again for the details, this is bring more clarity on your use case and expectations.

As per the expected behaviour defined in this issue:

> Terraform should have been able to execute an update to a slack notification without passing in a token. In the same way that the UI populates that for the user, it would be super if the backend was capable of retrieving the token itself
> 

The `notifier_id` is effectively solving this specific use case, being able to **update** an existing alert configuration notification without having to provide the original api key.

With this in mind, seeing both of your inputs:

> hopefully this provides some insight into the true outcome I had imagined (Manually configured Slack OAuth integration in the UI, with the ability to reference that within Terraform in order to create new alert resources)
> 

> It's about creating new alerts while Slack integration already created via web UI.
> 

It is clear that there is different use case that has not been addressed. The end goal would be to provide a mechanism that enables using an existing atlas third party integration when using the `mongodbatlas_alert_configuration` resource to **create** or **update** a notification. This is relevant for the Slack integration as it has to be created through the Atlas UI using OAuth2 (as described in [API docs](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Third-Party-Integrations/operation/createThirdPartyIntegration)). This would require a change from the Atlas API to either enable obtaining the raw value of the api key defined in a third party integration, or allowing to reference the third party integration when creating an alert configuration.

@Zuhairahmed's input will be valuable here in case there is any existing work or alternative I am not aware of. If this is in fact a valid feature that is not being worked on, I will follow up here so that we create a separate ticket to track this. Please let me know if you have any additional points or concerns to add.

**Comment by Zuhairahmed** [comment]
@nthienan / @TheQueenIsDead Terraform support for Slack Third Party Integration via OAuth2 is on our roadmap, but likely several quarters away. Internal Jira ticket: [CLOUDP-215081](https://jira.mongodb.org/browse/CLOUDP-215081) 

For tracking purposes since this Issue is closed, suggest @AgustinBettati do you want to open a new GitHub issue and link to this one? This way folks can follow open GitHub Issue and we can update that thread when this feature has been released

**Comment by AgustinBettati** [comment]
In our feature request website we have an existing post for the support of slack third party integration: https://feedback.mongodb.com/forums/924145-atlas/suggestions/45030598-support-slack-in-third-party-integration. 
In addition, I have created a post for being able to reference existing third party integrations from alert configuration resource: https://feedback.mongodb.com/forums/924145-atlas/suggestions/47533295-reference-existing-third-party-integration-in-aler.
